crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -2515,3 +2518,93 @@ fn range_size_hint(range: &impl RangeBounds<TxNumber>) -> Option<usize> {
     };
     end.checked_sub(start).map(|x| x as _)
 }
+
+/// Limits on how long pruning can run before it's forced to stop and thereby yield the
+/// [`DatabaseProviderRW`] hook.
+#[derive(Debug, Clone, Copy)]
+pub struct PruneLimiter {
+    /// Maximum units to delete from the database per block.
+    deleted_units_limit: Option<usize>,
+    /// Current number of units (entries) that have been deleted from database during the prune
+    /// job.
+    deleted_units_count: usize,
+    /// The max time one prune job can run.
+    job_timeout: Option<Duration>,
+    /// Time at which the prune job was started.
+    start: Instant,
+    /// Prune job has timed out.
+    timed_out: bool,
+}
+
+impl PruneLimiter {
+    /// Returns a new instance.
+    pub const fn new(
+        deleted_units_limit: Option<usize>,
+        job_timeout: Option<Duration>,
+        start: Instant,
+    ) -> Self {
+        Self { deleted_units_limit, deleted_units_count: 0, job_timeout, start, timed_out: false }
+    }
+
+    /// Returns a new instance with a fraction of the limit on units of the old instance.
+    pub fn new_with_fraction_of_units_limit(old: Self, denominator: NonZeroUsize) -> Self {
+        let Self { deleted_units_limit, job_timeout, start, .. } = old;
+
+        Self::new(deleted_units_limit.map(|limit| limit / denominator), job_timeout, start)
```

Comment:
how do you feel about making it a builder pattern, so it'd be a `with_fraction_of_units_limit(mut self, fraction: NonZero) -> Self` to just replace one field?

Reply:
builder pattern should be reserved for builders, so added a builder now, took some time but looks clean